File: lua/tests/scripts/python/tree.py
# Creates a binary tree of depth 'd'
def make_tree(d):

    if d > 0:
        d -= 1
        return (make_tree(d), make_tree(d))
    return (None, None)


def check_tree(node):

    (l, r) = node
    if l is None:
        return 1
    else:
        return 1 + check_tree(l) + check_tree(r)


def tree(n, min_depth=4):

        if (n < 0):
                return 0

        max_depth = max(min_depth + 2, n)
        stretch_depth = max_depth + 1

        t = make_tree(stretch_depth)
        check_tree(t)

        long_lived_tree = make_tree(max_depth)

        mmd = max_depth + min_depth
        for d in range(min_depth, stretch_depth, 2):
                iterations = 2 ** (mmd - d)
                cs = 0
                for i in range(iterations):
                        cs = cs + check_tree(make_tree(d))

        check_tree(long_lived_tree)

        return 0

File: core/tests/scripts/lua/tree.lua
local function BottomUpTree(depth)
        if depth > 0 then
                depth = depth - 1
                local left, right = BottomUpTree(depth), BottomUpTree(depth)
                return {left, right}
        else
                return {}
        end
end

local function ItemCheck(tree)
        if tree[1] then
                return 1 + ItemCheck(tree[1]) + ItemCheck(tree[2])
        else
                return 1
        end
end

local function tree(N)
        if (N < 0) then
                return 0
        end

        local mindepth = 4
        local maxdepth = mindepth + 2
        if maxdepth < N then
                maxdepth = N
        end

        local stretchdepth = maxdepth + 1
        local stretchtree = BottomUpTree(stretchdepth)
        ItemCheck(stretchtree)

        local longlivedtree = BottomUpTree(maxdepth)
        local mmd = maxdepth + mindepth

        for depth = mindepth, maxdepth, 2 do
                local iterations = 2 ^ (mmd - depth)
                local check = 0
                for i = 1, iterations do
                        check = check + ItemCheck(BottomUpTree(depth))
                end
        end
        ItemCheck(longlivedtree)

        return 0
end

function Test(N)
        return tree(N)
end
